dragonfly/block/trapdoor.go

```diff
@@ -0,0 +1,135 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/dragonfly/block/wood"
+	"github.com/df-mc/dragonfly/dragonfly/entity/physics"
+	"github.com/df-mc/dragonfly/dragonfly/item"
+	"github.com/df-mc/dragonfly/dragonfly/world"
+	"github.com/go-gl/mathgl/mgl64"
+	"math"
+)
+
+// Trapdoor is a block that can be used as an openable 1x1 barrier
+type Trapdoor struct {
+	noNBT
+
+	// Wood is the type of wood of the trapdoor. This field must have one of the values found in the material
+	// package.
+	Wood wood.Wood
+	// Facing is the direction the trapdoor is facing
+	Facing world.Direction
+	// Open is whether or not the trapdoor is open
+	Open bool
+	// Top is whether the trapdoor occupies the top or bottom part of a block
+	Top bool
+}
+
+// AABB ...
+func (t Trapdoor) AABB(pos world.BlockPos, w *world.World) []physics.AABB {
+	if t.Open {
+		return []physics.AABB{physics.NewAABB(mgl64.Vec3{}, mgl64.Vec3{1, 1, 1}).ExtendTowards(int(t.Facing.Face()), -0.8125)}
+	}
+	if t.Top {
+		return []physics.AABB{physics.NewAABB(mgl64.Vec3{}, mgl64.Vec3{1, 1, 1}).ExtendTowards(int(world.FaceDown), -0.8125)}
+	}
+	return []physics.AABB{physics.NewAABB(mgl64.Vec3{}, mgl64.Vec3{1, 1, 1}).ExtendTowards(int(world.FaceUp), -0.8125)}
+}
+
+// UseOnBlock handles the directional placing of trapdoors and makes sure they are properly placed upside down
+// when needed.
+func (t Trapdoor) UseOnBlock(pos world.BlockPos, face world.Face, clickPos mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) bool {
+	pos, face, used := firstReplaceable(w, pos, face, t)
+	if !used {
+		return false
+	}
+	t.Facing = user.Facing().Opposite()
+	t.Top = (clickPos.Y() > 0.5 && face != world.FaceUp) || face == world.FaceDown
+
+	place(w, pos, t, user, ctx)
+	return placed(ctx)
+}
+
+// Activate ...
+func (t Trapdoor) Activate(pos world.BlockPos, clickedFace world.Face, w *world.World, u item.User) {
+	t.Open = !t.Open
+	w.SetBlock(pos, t)
+}
+
+// BreakInfo ...
+func (t Trapdoor) BreakInfo() BreakInfo {
+	return BreakInfo{
+		Hardness:    3,
+		Harvestable: alwaysHarvestable,
+		Effective:   axeEffective,
+		Drops:       simpleDrops(item.NewStack(t, 1)),
+	}
+}
+
+// CanDisplace ...
+func (t Trapdoor) CanDisplace(l world.Liquid) bool {
```

Comment:
For the block to be waterloggable, the `world.LiquidDisplacer` interface has to be implemented, which also includes the `SideClosed` method. (Which I think should always return false in this case.)